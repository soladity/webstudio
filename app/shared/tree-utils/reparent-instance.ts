import produce from "immer";
import { type Instance } from "@webstudio-is/sdk";
import { findInstanceById } from "./find-instance";
import { deleteInstanceMutable } from "./delete-instance";
import {
  insertInstanceMutable,
  type InstanceInsertionSpec,
} from "./insert-instance";

export type InstanceReparentingSpec = {
  id: Instance["id"];
  parentId: Instance["id"];
  position: InstanceInsertionSpec["position"];
};

export const reparentInstance = produce(
  (draftRootInstance: Instance, spec: InstanceReparentingSpec) => {
    const instance = findInstanceById(draftRootInstance, spec.id);
    if (instance === undefined) return;
    deleteInstanceMutable(draftRootInstance, instance.id);
    insertInstanceMutable(draftRootInstance, instance, spec);
  }
);
