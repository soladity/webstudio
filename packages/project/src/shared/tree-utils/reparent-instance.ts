import type { Instance } from "@webstudio-is/project-build";
import { findInstanceById } from "./find-instance";
import { findParentInstance } from "./find-parent-instance";
import { deleteInstanceMutable } from "./delete-instance";
import { insertInstanceMutable } from "./insert-instance";

export const reparentInstanceMutable = (
  rootInstance: Instance,
  instanceId: Instance["id"],
  newParentId: Instance["id"],
  newPosition: number | "end"
) => {
  const instance = findInstanceById(rootInstance, instanceId);
  if (instance === undefined) {
    return;
  }

  if (newPosition === "end") {
    deleteInstanceMutable(rootInstance, instance.id);
    insertInstanceMutable(rootInstance, instance, {
      parentId: newParentId,
      position: "end",
    });
    return;
  }

  let newPositionAdjusted = newPosition;

  // If parent is the same, we need to adjust the position to account for the removal of the instance.
  const currentParent = findParentInstance(rootInstance, instance.id);
  if (currentParent !== undefined && currentParent.id === newParentId) {
    const currentPosition = currentParent.children.findIndex(
      (child) => child.type === "instance" && child.id === instance.id
    );
    if (currentPosition < newPosition) {
      newPositionAdjusted--;
    }
  }

  deleteInstanceMutable(rootInstance, instance.id);
  insertInstanceMutable(rootInstance, instance, {
    parentId: newParentId,
    position: newPositionAdjusted,
  });
};
