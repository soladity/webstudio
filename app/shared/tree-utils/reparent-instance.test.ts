import { type Instance } from "@webstudio-is/sdk";
import { type InstanceReparentingSpec } from ".";
import { reparentInstance } from "./reparent-instance";

describe("Reparent instance", () => {
  test(`div 1 > div 2, div 3  => div 1 > div 2 > div 3`, () => {
    const rootInstance: Instance = {
      component: "Box",
      id: "1",
      style: {},
      children: [
        {
          component: "Box",
          id: "2",
          style: {},
          children: [],
        },
        {
          component: "Box",
          id: "3",
          style: {},
          children: [],
        },
      ],
    };
    const instanceReparentingSpec: InstanceReparentingSpec = {
      parentId: "2",
      position: "end",
      id: "3",
    };
    const result = reparentInstance(rootInstance, instanceReparentingSpec);
    expect(result).toMatchSnapshot();
  });
});
