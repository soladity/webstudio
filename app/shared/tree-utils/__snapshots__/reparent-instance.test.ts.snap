// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reparent instance div 1 > div 2, div 3  => div 1 > div 2 > div 3 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "component": "Box",
          "id": "3",
          "style": Object {},
        },
      ],
      "component": "Box",
      "id": "2",
      "style": Object {},
    },
  ],
  "component": "Box",
  "id": "1",
  "style": Object {},
}
`;
